File: autocontext/src/autocontext/integrations/anthropic/_taxonomy.py
"""Exception → reason-key lookup for Anthropic SDK exceptions."""
from __future__ import annotations

import anthropic

from autocontext.production_traces.taxonomy import (
    ANTHROPIC_ERROR_REASONS,
    AnthropicErrorReasonKey,
)


def map_exception_to_reason(exc: BaseException) -> AnthropicErrorReasonKey:
    name = type(exc).__name__
    return ANTHROPIC_ERROR_REASONS.get(name, "uncategorized")  # type: ignore[return-value]


def is_mapped_class_present(class_name: str) -> bool:
    """Return True if class_name is accessible from the anthropic package (top-level or _exceptions)."""
    if hasattr(anthropic, class_name):
        return True
    try:
        from anthropic import _exceptions  # noqa: PLC0415
        return hasattr(_exceptions, class_name)
    except ImportError:
        return False

File: ts/src/integrations/anthropic/taxonomy.ts
/**
 * Exception → reason-key lookup for Anthropic errors.
 * Uses constructor.name to map SDK error class names to taxonomy keys.
 * Mirror of Python _taxonomy.py for the Anthropic provider.
 */
import {
  ANTHROPIC_ERROR_REASONS,
  type AnthropicErrorReasonKey,
} from "../../production-traces/taxonomy/anthropic-error-reasons.js";

/**
 * Look up err's class name in the Anthropic taxonomy.
 * Returns "uncategorized" on miss or if exc is not an object.
 */
export function mapExceptionToReason(exc: unknown): AnthropicErrorReasonKey {
  if (exc == null || typeof exc !== "object") return "uncategorized";
  const name = (exc as { constructor?: { name?: string } }).constructor?.name;
  if (!name) return "uncategorized";
  return (ANTHROPIC_ERROR_REASONS[name] ?? "uncategorized") as AnthropicErrorReasonKey;
}
